### ISSUE TITLE
Sphinx-copybutton is not visible until hovered or focused

### ISSUE THREAD

**Issue by gabalafou** [Quansight-Labs/czi-scientific-python-mgmt#78]

**Comment by gabalafou** [comment]
I want to leave this one up to design whether we should change the user experience here, and if so, how.

**Comment by Carreau** [comment]
I'm going to have what might be a controversial opinion, but IIRC the goal of sphinx-copy-button was to simplify copying when `>>>` and `...` were present in the cell and thus annoying to remove for large block of code when you actually did not wanted to copy "all".

With css [`user-select`](https://developer.mozilla.org/en-US/docs/Web/CSS/user-select) we've already made the prompt and other line numbers non-selectionable, so there should be fewer, if no need at all for sphinx copy-button.

I think there is a better way to achieve this than the copy button by using already existing user interface. I think we likely are looking for something like `user-select:contain`, or `user-select:all`.

Right now you still have to drag to select a full cell.
Maybe making a quadruple click in a cell select the whole text (triple click is select the whole line by default)

```
        document.querySelectorAll('pre').forEach(pre => {
            pre.addEventListener('click', (function() {
                let clickCount = 0;
                let timer;

                return function(event) {
                    clickCount++;
                    clearTimeout(timer);

                    if (clickCount === 4) {
                        let range = document.createRange();
                        range.selectNodeContents(this);
                        let selection = window.getSelection();
                        selection.removeAllRanges();
                        selection.addRange(range);
                        clickCount = 0; // Reset the count after quadruple click
                    }

                    timer = setTimeout(() => clickCount = 0, 400); // Reset click count after 400ms
                };
            })());
        });
```

Does the trick.

Now for accessibility, I believe we "just" need the copy action of the browser to copy the content of the cell when focused.

And with that we'll just have no _need_ for sphinx copy button, and then decide if we want to phase it out.

**Comment by Carreau** [comment]
> Does the Copy Button Stay or Go?: From an accessibility perspective, I think it is important to keep the copy button. It provides a simple and clear one-click action for copying code, which is useful for users with motor impairments or those using assistive tech. The button will ensure that all users have an accessible and straightforward way to copy code.

You seem to have missed :

> Now for accessibility, I believe we "just" need the copy action of the browser to copy the content of the cell when focused.

If the cell is focusable and has tabindex=..., and when focused the native 'copy' function works be copying the full content, why would we need an extra copy button ? I mean from an accessibility perspective wouldn't `edit > copy`, or `contextual menu > copy` be better alternatives ? Plus keyboard shortcuts would work as well. 

I don't think it is easier for a user with impairment to managed to either point the mouse as a small target, or use keyboard navigation to move away from the current selection to find the copy button. That at least my experience with people I know have accessibility issues are more likely to use keyboard shortcuts.

Likely i'm miss-understanding how assistive tech might work, I'm representing something like keyboard like navigation where you focus each item in turn, and so would likely have already a code block focused and then want to "copy" it's content.

I'm also curious about what it give as screen readers, having copy buttons mean that while navigating the page every other other will read "copy button" which seem to me like a worse experience.

Now I agree that it's more *discoverable* and easier for some categories of users, mostly *novices* that don't know keyboard shortcuts, or contextual menus. I can also be convinced it's 
it's more *convenient* for typical users without accessibility issues that can easily reach for the copy-button. But I don't believe that this copy button is better for people with accessibility issues in general.


One thing I'm afraid is wether we are just trying to improve the copy button **because it is there**, which is a sunk-cost fallacy. Instead of asking "if there was no copy-button what would we do ?". I'm also aware of [chesterton's fence](https://fs.blog/chestertons-fence/), but I believe in this case we clearly know why sphinx copy-button was added : to allow copying without getting the prompts in the clipboard – reason which definitely does ont exists anymore as the prompt are not selectable.

**Comment by Carreau** [comment]
On the if we keep it how to improve it.:

When the code block has a header bar I think the button could/should be in the header bar and always be visible. For example in a case like this:


<img width="742" alt="Screenshot 2024-06-26 at 14 13 45" src="https://github.com/Quansight-Labs/czi-scientific-python-mgmt/assets/335567/ba6f4305-d8bf-4007-b9b7-c42678d40e3f">

I think a [clipboard icon](https://fontawesome.com/v4/icon/clipboard) would be also more recognizable.


Another question if we keep the button, why have it only on multiline code blocks ? I'm assuming pre tags (and even whole paragraphs) also deserved to have copy button ? like in:

> banner by prepending the following HTML to all pages (be sure to replace `URL_OF_STABLE_VERSION_OF_PROJECT` with a valid URL, and adjust styling as desired):

I can see how `URL_OF_STABLE_VERSION_OF_PROJECT` could be a copy target.

**Comment by gabalafou** [comment]
I like these arguments for getting rid of the extension. However, even if we get rid of it in the PyData Theme docs, there's still the question of whether theme adopters will still put the extension on their site **even if** we make the changes @Carreau suggests. 

My understanding is that part of the value proposition of the PyData Theme is that it plays nicely with certain extensions and even carries over some desirable properties (such as more accessible color contrasts) to some of these extensions (for example, Sphinx Design).

In the short term, we may want to make a few changes that only apply to the copy button extension when it's used in conjunction with the theme. In the long term, we may want to work with the extension authors to implement accessibility best practices.

**Comment by gabalafou** [comment]
Alternatively, we could take a stronger stance and tell theme adopters that they shouldn't use the copy button extension.

**Comment by Carreau** [comment]
I don't think we should take a strong stance right now. I would much better explore what we can do without, and do a 3 step removal: 
 - opt-out
 - opt-in
 - remove compatibility

This let us explore wether the copy button is actually (un)necessary.

**Comment by gabalafou** [comment]
> 3 step removal

Love that

**Comment by Carreau** [comment]
With the step-0 improve the theme html/js to make copy button unnecessary :-)

**Comment by gabalafou** [comment]
Could I get an update on this? 

My understanding is two-fold:

1. We will make the copy button always visible in the short term
2. Over the long term we will climb @Carreau's three-step removal ladder, evaluating at each rung whether we move to the next.

Is that correct? 

If so, we can move this issue to dev and call it done when the button is always visible. We should open a separate issue to track removing the extension.

**Comment by Carreau** [comment]
I'll add myself as assignee and will look into making it always visible.

**Comment by smeragoel** [comment]
@gabalafou yes your summary is correct, that is the plan

**Comment by gabalafou** [comment]
Bump on this. If there isn't current work underway on this, could I take it over?

**Comment by Carreau** [comment]
Switched the assignee from me to you.
